### ISSUE TITLE
Invalid use of delete [] in gengraph

### ISSUE THREAD

**Issue by szhorvat** [igraph/igraph#2169]
**Describe the bug**

I am getting the following warning from GCC 11.3 when using LTO:

```
In member function ‘__dt_base ’,
    inlined from ‘__ct ’ at /home/szhorvat/Repos/igraph/src/games/degree_sequence_vl/gengraph_graph_molloy_hash.cpp:123:1,
    inlined from ‘igraph_degree_sequence_game_vl’ at /home/szhorvat/Repos/igraph/src/games/degree_sequence_vl/gengraph_mr-connected.cpp:143:9:
/home/szhorvat/Repos/igraph/src/games/degree_sequence_vl/gengraph_degree_sequence.cpp:53:18: warning: ‘operator delete []’ called on pointer ‘_672’ with nonzero offset 16 [-Wfree-nonheap-object]
   53 |         delete[] deg;
      |                  ^
/home/szhorvat/Repos/igraph/src/games/degree_sequence_vl/gengraph_degree_sequence.cpp: In function ‘igraph_degree_sequence_game_vl’:
/home/szhorvat/Repos/igraph/src/games/degree_sequence_vl/gengraph_graph_molloy_optimized.cpp:171:62: note: returned from ‘operator new []’
  171 |     igraph_integer_t *hc = new igraph_integer_t[2 + n + a / 2]; // to store n,a,deg[] and links[]
      |                                                              ^
```

I believe this comes from the constructor `degree_sequence(igraph_integer_t n0, igraph_integer_t *degs)` which is called here: https://github.com/igraph/igraph/blob/a5661a3334a3f1722edd99f2df385529e301bf66/src/games/degree_sequence_vl/gengraph_graph_molloy_optimized.cpp#L129 using an address that was not allocated with `new []`.

**Version information**
`develop`, 0.10-RC2

**Comment by ntamas** [comment]
Yes, that's _formally_ incorrect, but the compiler fails to recognize that we call `dd.detach()` before `dd` goes out of context, so by the time we would reach the offending `delete[] deg` call, `deg` was already invalidated by `.detach()` so we would never call `delete[]`.

It would be great if we could somehow make this clear to the linker, though.

**Comment by ntamas** [comment]
I'll experiment with rewriting `degree_sequence` a bit. It's a bit weird for me that this class essentially takes ownership of the pointer that was passed into it and attempts to free it later with `delete[]` without knowing how it was allocated in the first place.

**Comment by szhorvat** [comment]
Ah, you're right. I suggest we don't bother then. It's just a warning, and appears only with LTO. It doesn't trigger an error, even with warnings-as-errors.

**Comment by szhorvat** [comment]
P.S. At some point I'll probably throw out the gengraph code and rewrite the algorithm from scratch.

**Comment by ntamas** [comment]
The ownership issues of `degree_sequence` were improved a bit in 298ab92f7; hopefully this version of the code does not trip up the sanitizer any more.

**Comment by szhorvat** [comment]
It was not the sanitizer. Sanitizers do runtime checks and can avoid false positives like this. This was a compile-time warning from GCC when using LTO, so it can check code across compilation unit boundaries.

Yes, the compiler is now happy. Thanks!

**Comment by szhorvat** [comment]
P.S. We still get a bunch of annoying deprecation warnings from GCC, but that's really GCC's fault for warning about use of deprecated functions from other deprecated functions. There is no reasonable solution to this. Hopefully GCC will catch up with Clang.
